core/src/main/java/org/apache/iceberg/BasePartitionStatistics.java

```diff
@@ -0,0 +1,353 @@
+        break;
+      case EQUALITY_DELETES:
+        this.equalityDeleteRecordCount += file.recordCount();
+        this.equalityDeleteFileCount += 1;
+        break;
+      default:
+        throw new UnsupportedOperationException("Unsupported file content type: " + file.content());
+    }
+
+    if (snapshot != null) {
+      updateSnapshotInfo(snapshot.snapshotId(), snapshot.timestampMillis());
+    }
+
+    // Note: Not computing the `TOTAL_RECORD_COUNT` for now as it needs scanning the data.
+  }
+
+  /**
+   * Updates the modified time and snapshot ID for the deleted manifest entry.
+   *
+   * @param snapshot the snapshot corresponding to the deleted manifest entry.
+   */
+  void deletedEntry(Snapshot snapshot) {
+    if (snapshot != null) {
+      updateSnapshotInfo(snapshot.snapshotId(), snapshot.timestampMillis());
+    }
+  }
+
+  /**
+   * Decrement the counters as it was included in the previous stats and updates the modified time
+   * and snapshot ID for the deleted manifest entry.
+   *
+   * @param snapshot the snapshot corresponding to the deleted manifest entry.
+   */
+  void deletedEntryForIncrementalCompute(ContentFile<?> file, Snapshot snapshot) {
+    Preconditions.checkArgument(file.specId() == specId, "Spec IDs must match");
+
+    switch (file.content()) {
+      case DATA:
+        this.dataRecordCount -= file.recordCount();
+        this.dataFileCount -= 1;
+        this.totalDataFileSizeInBytes -= file.fileSizeInBytes();
+        break;
+      case POSITION_DELETES:
+        this.positionDeleteRecordCount -= file.recordCount();
+        if (file.format() == FileFormat.PUFFIN) {
+          this.dvCount -= 1;
+        } else {
+          this.positionDeleteFileCount -= 1;
+        }
+
+        break;
+      case EQUALITY_DELETES:
+        this.equalityDeleteRecordCount -= file.recordCount();
+        this.equalityDeleteFileCount -= 1;
+        break;
+      default:
+        throw new UnsupportedOperationException("Unsupported file content type: " + file.content());
+    }
+
+    if (snapshot != null) {
+      updateSnapshotInfo(snapshot.snapshotId(), snapshot.timestampMillis());
+    }
+  }
+
+  /**
+   * Appends statistics from given entry to current entry.
+   *
+   * @param entry the entry from which statistics will be sourced.
+   */
+  void appendStats(PartitionStatistics entry) {
+    Preconditions.checkArgument(entry.specId() != null, "Invalid spec ID: mull");
+    Preconditions.checkArgument(entry.specId().equals(this.specId), "Spec IDs must match");
+
+    this.dataRecordCount += entry.dataRecordCount();
+    this.dataFileCount += entry.dataFileCount();
+    this.totalDataFileSizeInBytes += entry.totalDataFileSizeInBytes();
+    this.positionDeleteRecordCount += entry.positionDeleteRecordCount();
+    this.positionDeleteFileCount += entry.positionDeleteFileCount();
+    this.equalityDeleteRecordCount += entry.equalityDeleteRecordCount();
+    this.equalityDeleteFileCount += entry.equalityDeleteFileCount();
```

Comment:
What happens when one of these are null?

Reply:
They can't be null, because this is on the write path where we use the full V2 or V3 schema for read/write. Added a comment